site/content/blog/2020-09-07-improving-performance-through-load-testing.md

```diff
@@ -0,0 +1,83 @@
+canonicalUrl: https://mattermost.com/blog/improving-performance-through-load-testing/
+---
+
+Have you ever wondered how many active users your application can handle at the same time? If so, you're not alone. Here at Mattermost we're building a highly concurrent messaging platform for team collaboration that needs to potentially serve up to several thousands of users simultaneously. 
+
+While standard functional testing (e.g. unit tests) is critical to verify correct behavior of your application, it's usually not sufficient to guarantee its performance at scale. This is where some form of performance testing can be particularly useful. Benchmarks do a good job at measuring the performance impact of single functions or limited portions of a codebase, but they can't capture how an application is going to perform as a whole under real-world conditions. To achieve this you need **load testing**.
+
+## What is load testing?
+
+Load testing is the process of applying a certain amount of load to a target system in order to understand how it behaves.
+The more realistic the generated load is, the more accurate and insightful the final results will be.
+
+## Why does it matter?
+
+Load testing can be used to improve both the performance and the overall quality of your application by helping you to:
+
+- Identify performance bottlenecks in your system.
+- Determine the overall capacity of your application.
+- Prevent performance regressions between releases.
+- Discover insidious bugs that passed functional testing.
+
+As an example, these are some of non-performance related, but rather significant issues we've been able to detect while load testing:
+
+- Data races
+- Application crashes
+- Deadlocks
+- Memory leaks
+
+As you can see, running load tests can guard your software against problems of different nature. By simulating users you get a chance to realistically test a system before it hits production and potentially avoid several issues down the line.
+
+## How are we doing it?
+
+There are plenty of extremely valid load testing frameworks available; some of the best are free and open source (JMeter, Gatling, k6 to name a few). In our case, after a thorough examination, we decided to develop a custom, in-house tool. We accepted trading implementation time and complexity for improved simulation and finer control over the entire process. We've recently released [version 1.0](https://github.com/mattermost/mattermost-load-test-ng/releases/tag/v1.0.0) and we're quite happy with the choice we made. While the tool itself is not designed for general purpose use, it can still serve as a good example of how we tackled the problem of designing and developing a custom load testing engine.
+
+### Simulating the load
+
+As the definition says, the core of the process is load generation. However, in order for results to be meaningful, we need the software to mimic actual user behavior. We achieve this by leveraging data coming from real users through the use of:
```

Comment:
Are there any other canonical load generation methods? If so, we can mention them and explain why we picked hitting directly to the endpoints by using a user client.

Reply:
I think what we are doing is not far off from the *canonical* way of load testing. There are other strategies of course like relaying on some pre-recorded users scenarios instead of actually building the simulation from scratch.

I think the most significant difference that our tool introduces is the use of a feedback loop to converge and I've attempted to explain from an high level point of view why and how we did so.